engine/src/main/java/org/camunda/bpm/engine/impl/persistence/entity/util/TypedValueField.java

```diff
@@ -66,34 +66,43 @@
   public TypedValueField(ValueFields valueFields, boolean notifyOnImplicitUpdates) {
     this.valueFields = valueFields;
     this.notifyOnImplicitUpdates = notifyOnImplicitUpdates;
-    this.updateListeners = new ArrayList<TypedValueUpdateListener>();
+    this.updateListeners = new ArrayList<>();
   }
 
   public Object getValue() {
-    TypedValue typedValue = getTypedValue();
+    TypedValue typedValue = getTypedValue(false);
     if (typedValue != null) {
       return typedValue.getValue();
     } else {
       return null;
     }
   }
 
-  public TypedValue getTypedValue() {
-    return getTypedValue(true);
+  public TypedValue getTypedValue(boolean asTransientValue) {
+    return getTypedValue(true, asTransientValue);
   }
 
-  public TypedValue getTypedValue(boolean deserializeValue) {
-    if (cachedValue != null && cachedValue instanceof SerializableValue && Context.getCommandContext() != null) {
-      SerializableValue serializableValue = (SerializableValue) cachedValue;
-      if(deserializeValue && !serializableValue.isDeserialized()) {
-        // clear cached value in case it is not deserialized and user requests deserialized value
+  public TypedValue getTypedValue(boolean deserializeValue, boolean asTransientValue) {
+    if (Context.getCommandContext() != null) {
+      // in some circumstances we must invalidate the cached value instead of returning it
+
+      if (cachedValue != null && cachedValue instanceof SerializableValue) {
+        SerializableValue serializableValue = (SerializableValue) cachedValue;
+        if(deserializeValue && !serializableValue.isDeserialized()) {
+          // clear cached value in case it is not deserialized and user requests deserialized value
+          cachedValue = null;
+        }
+      }
+
+      if (cachedValue != null && ((asTransientValue && !cachedValue.isTransient()) || !asTransientValue && cachedValue.isTransient())) {
```

Comment:
Can we use XOR here? If both are the same, it will be `false`, otherwise it will be `true` and we clear the cache.
```suggestion
      if (cachedValue != null && (asTransientValue ^ cachedValue.isTransient()) {
```